Live migration with attached storage volumes (from Incus)

markylaing opened the pull request:
Enables live-migration of an instance with volumes attached as long as those disks are also on a remote storage pool.

Created as draft while I finish writing tests.

Closes #12694

tomponline commented on lxd/device/disk.go:
```diff
@@ -500,7 +500,7 @@ func (d *disk) validateConfig(instConf instance.ConfigReader) error {
 
-				if remoteInstance != nil {
+				if remoteInstance != nil && remoteInstance.ID != instConf.ID() {
 					return fmt.Errorf("Custom volume is already attached to an instance on a different node")
```
Please can you change the "node" reference here to "cluster member" in a separate commit.

tomponline commented on lxd/instance/drivers/driver_qemu.go:
```diff
@@ -6857,6 +6857,39 @@ func (d *qemu) migrateSendLive(pool storagePools.Pool, clusterMoveSourceName str
 		return err
 	}
 
```
Please remove the "incus issue" reference in the commit message as not relevant.

tomponline commented on lxd/instance/drivers/driver_qemu.go:
```diff
@@ -7338,6 +7371,39 @@ func (d *qemu) MigrateReceive(args instance.MigrateReceiveArgs) error {
+			vol := diskPool.GetVolume(storageDrivers.VolumeTypeCustom, storageDrivers.ContentTypeBlock, project.StorageVolume(storageProjectName, dev.Config["source"]), nil)
+			volCopy := storageDrivers.NewVolumeCopy(vol)
+
+			// Call MigrateVolume on the source.
```
Is this comment correct?

tomponline commented on lxd/instance/drivers/driver_qemu.go:
```diff
@@ -6857,6 +6857,39 @@ func (d *qemu) migrateSendLive(pool storagePools.Pool, clusterMoveSourceName str
+		}
+
+		// Load the pool for the disk.
+		diskPool, err := storagePools.LoadByName(d.state, dev.Config["pool"])
```
An optimisation here would be to store the diskPool in a map keyed by name, so that if there are multiple disks from the same pool you dont have to load the pool from DB multiple times.

markylaing commented:
I've addressed the comments and the test (https://github.com/canonical/lxd-ci/pull/250) is passing for me locally. It is also failing at the expected point in github actions, so I believe it will pass once this is merged and in latest/edge.
